tracing-journald/tests/journal.rs

```diff
@@ -0,0 +1,83 @@
+#![cfg(target_os = "linux")]
+
+use std::collections::HashMap;
+use std::process::Command;
+
+use tracing::{error, info, warn};
+use tracing_journald::Subscriber;
+use tracing_subscriber::subscribe::CollectExt;
+use tracing_subscriber::Registry;
+
+fn with_journald(f: impl FnOnce()) {
+    let sub = Registry::default().with(Subscriber::new().unwrap().with_field_prefix(None));
+    tracing::collect::with_default(sub, f);
+}
+
+/// Read from journal with `journalctl`.
+///
+/// `test_name` is the randomized name of the test being run, and gets
+/// added as `TEST_NAME` match to the `journalctl` call, to make sure to
+/// only select journal entries originating from and relevant to the
+/// current test.
+fn read_from_journal(test_name: &str) -> Vec<HashMap<String, String>> {
```

Comment:
we should probably handle the case where `journalctl` is not found, and have the tests return early if it's not present, rather than failing. the `#![cfg(target_os = "linux")]` prevents them from failing on macOS/BSD/Windows/etc, but we might also be on a Linux distro that uses some other init system, or in some kind of environment where the `journalctl` executable is not available, like in a docker container.

for example, these tests fail on my laptop running Void Linux, which uses `runit` rather than `systemd`. checking for the presence of a `journalctl` executable seems like a pretty good way to determine whether we're in an environment where it makes sense to test the `journald` subscriber.

Reply:
we could probably do this by running `which journalctl` beforehand, or by looking at the exit code from running the journalctl command.

Reply:
Another option would be to short-circuit the tests if `Subscriber::new()` fails, but I think that would be a little less robust to weird environments.

Reply:
Alright, I'll try to safely cancel the tests if `journalctl` fails to execute with `ENOENT` (I guess that's the simplest approach), but I'd need you to test this, because I'm not running any systems without systemd, and I do have to admit that I don't care enough to install something like Void somewhere just for this issue :innocent:

Reply:
Yup, that's fair --- happy to test it out.

I'm actually curious to learn if journalctl is available in the containers that GitHub Actions will run these tests in, too...

Reply:
@hawkw It is, at least in the standard ones.  I wrote these tests for libsystemd-rs in the past, and simply copied and adapted them.  I must admit that I wouldn't have offered to write these tests had I not known beforehand that they'd work :innocent:

Reply:
@hawkw I pushed the change.

I haven't found a way to properly skip tests at runtime in Rust's rather basic test runner, and I didn't want to go all the way to a custom harness, so in want of any better idea I made the tests print a warning to stderr in case `journalctl --version` doesn't work.

It's not pretty, but I hope it's good enough :innocent:

Reply:
@lunaryorn 
> I haven't found a way to properly skip tests at runtime in Rust's rather basic test runner,

I would just have the tests return early. :)